src/Processors/Transforms/DeduplicationTokenTransforms.h

```diff
@@ -0,0 +1,236 @@
+#pragma once
+
+#include <Processors/Chunk.h>
+#include <Processors/ISimpleTransform.h>
+
+#include <base/defines.h>
+#include "Common/Logger.h"
+
+
+namespace DB
+{
+    class RestoreChunkInfosTransform : public ISimpleTransform
+    {
+    public:
+        RestoreChunkInfosTransform(Chunk::ChunkInfoCollection chunk_infos_, const Block & header_)
+                : ISimpleTransform(header_, header_, true)
+                , chunk_infos(std::move(chunk_infos_))
+        {}
+
+        String getName() const override { return "RestoreChunkInfosTransform"; }
+
+        void transform(Chunk & chunk) override;
+
+    private:
+        Chunk::ChunkInfoCollection chunk_infos;
+    };
+
+
+namespace DeduplicationToken
+{
+    class TokenInfo : public ChunkInfoCloneable<TokenInfo>
+    {
+    public:
+        TokenInfo() = default;
+        TokenInfo(const TokenInfo & other) = default;
+
+        String getToken() const;
+        String debugToken() const;
+
+        bool empty() const { return parts.empty(); }
+
+        bool isDefined() const { return stage == DEFINED; }
+
+        void addChunkHash(String part);
+        void defineSourceWithChunkHashes();
+
+        void setUserToken(const String & token);
+        void defineSourceWithUserToken(size_t block_number);
+
+        void setViewID(const String & id);
+        void defineViewID(size_t block_number);
+
+        void reset();
+
+    private:
+        String getTokenImpl() const;
+
+        void addTokenPart(String part);
+        size_t getTotalSize() const;
+
+        /* Token has to be prepared in a particular order.
+        * BuildingStage ensures that token is expanded according the foloving order.
+        * Firstly token is expanded with information about the source.
+        * It could be done with two ways: add several hash sums from the source chunks or provide user defined deduplication token and its sequentional block number.
+        *
+        * transition // method
+        * UNDEFINED -> DEFINE_SOURCE_WITH_HASHES // addChunkHash
+        * DEFINE_SOURCE_WITH_HASHES -> DEFINE_SOURCE_WITH_HASHES // addChunkHash
+        * DEFINE_SOURCE_WITH_HASHES -> DEFINED // defineSourceWithChankHashes
+        *
+        * transition // method
+        * UNDEFINED -> DEFINE_SOURCE_USER_TOKEN // setUserToken
+        * DEFINE_SOURCE_USER_TOKEN -> DEFINED // defineSourceWithUserToken
+        *
+        * After token is define it could be extended with view id and view block number. Actually it has to be expanded with view details if there is one or several views.
```

Comment:
```suggestion
        * After token is defined, it could be extended with view id and view block number. Actually it has to be expanded with view details if there is one or several views.
```